src/test/optimizer/strategy/dependent_group_by_reduction_rule_test.cpp

```diff
@@ -0,0 +1,324 @@
+  }
+
+  // Early out for LQP where table does not have a primary key/unique constraint
+  {
+    auto lqp =
+    PredicateNode::make(equals_(column_d_0, 17), stored_table_node_d);
+
+    const auto unmodified_lqp = apply_rule(rule, lqp);
+    EXPECT_EQ(lqp, unmodified_lqp);
+  }
+  // clang-format on
+}
+
+// Test that a removable column is removed when the single primary key column is present.
+// Check for the restored colum order.
+TEST_F(DependentGroupByReductionRuleTest, SingleKeyReduction) {
+  // clang-format off
+  auto lqp =
+  AggregateNode::make(expression_vector(column_a_0, column_a_1),
+                      expression_vector(sum_(column_a_0), sum_(column_a_1),
+                                        sum_(column_a_2)),
+                      stored_table_node_a);
+
+  const auto actual_lqp = apply_rule(rule, lqp);
+
+  const auto expected_lqp =
+  ProjectionNode::make(expression_vector(column_a_0, column_a_1, sum_(column_a_0), sum_(column_a_1), sum_(column_a_2)),
+    AggregateNode::make(expression_vector(column_a_0),
+                        expression_vector(sum_(column_a_0), sum_(column_a_1),
+                                          sum_(column_a_2), any_(column_a_1)),
+                        stored_table_node_a));
+  // clang-format on
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+// Test that a non-primary-key column is not removed if the full unique constraint is not present in the group by list.
+TEST_F(DependentGroupByReductionRuleTest, IncompleteKey) {
+  // clang-format off
+  auto lqp =
+  AggregateNode::make(expression_vector(column_b_0, column_b_2),
+                      expression_vector(sum_(column_b_0), sum_(column_b_1),
+                                        sum_(column_b_2)),
+                      stored_table_node_b);
+
+  const auto actual_lqp = apply_rule(rule, lqp);
+
+  const auto expected_lqp =
+  AggregateNode::make(expression_vector(column_b_0, column_b_2),
+                      expression_vector(sum_(column_b_0), sum_(column_b_1),
+                                        sum_(column_b_2)),
+                      stored_table_node_b);
+  // clang-format on
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+// Test that a group by with the full (multi-column) unique constraint is not altered.
+TEST_F(DependentGroupByReductionRuleTest, FullKeyGroupBy) {
+  // clang-format off
+  auto lqp =
+  AggregateNode::make(expression_vector(column_b_0, column_b_1),
+                      expression_vector(sum_(column_b_0), sum_(column_b_1),
+                                        sum_(column_b_2)),
+                      stored_table_node_b);
+
+  const auto actual_lqp = apply_rule(rule, lqp);
+
+  const auto expected_lqp =
+  AggregateNode::make(expression_vector(column_b_0, column_b_1),
+                      expression_vector(sum_(column_b_0), sum_(column_b_1),
+                                        sum_(column_b_2)),
+                      stored_table_node_b);
+  // clang-format on
```

Comment:
For LQPs that are not expected to change, compare against a deep_copy (Applies to whole file)

Reply:
Is 0070034  ok?

Reply:
I changed two tests.
In other tests, I am not comparing the original LQP against an optimised one but against a "new" lqp. That should be unproblematic, right?
But it's probably a dump idea to create the same LQP twice. So I'll change it to a `deep_copy()`'d one.

Reply:
Done.